cli_tools/gce_windows_upgrade/upgrader/workflows.go

```diff
@@ -0,0 +1,492 @@
+			&daisy.MachineImage{
+				MachineImage: computeBeta.MachineImage{
+					Name:           u.machineImageBackupName,
+					SourceInstance: u.InstanceURI,
+				},
+				Resource: daisy.Resource{
+					ExactName: true,
+					NoCleanup: true,
+				},
+			},
+		}
+		prevStep = stepBackupMachineImage
+	}
+
+	stepBackupOSDiskSnapshot, err := daisyutils.NewStep(w, "backup-os-disk-snapshot", prevStep)
+	if err != nil {
+		return err
+	}
+	stepBackupOSDiskSnapshot.CreateSnapshots = &daisy.CreateSnapshots{
+		&daisy.Snapshot{
+			Snapshot: compute.Snapshot{
+				Name:       u.osDiskSnapshotName,
+				SourceDisk: u.osDiskURI,
+			},
+			Resource: daisy.Resource{
+				ExactName: true,
+				NoCleanup: true,
+			},
+		},
+	}
+
+	stepCreateNewOSDisk, err := daisyutils.NewStep(w, "create-new-os-disk", stepBackupOSDiskSnapshot)
+	if err != nil {
+		return err
+	}
+	stepCreateNewOSDisk.CreateDisks = &daisy.CreateDisks{
+		&daisy.Disk{
+			Disk: compute.Disk{
+				Name:           u.newOSDiskName,
+				Zone:           u.zone,
+				Type:           u.osDiskType,
+				SourceSnapshot: u.osDiskSnapshotName,
+				Licenses:       []string{licenseToAdd[u.SourceOS]},
+			},
+			Resource: daisy.Resource{
+				ExactName: true,
+				NoCleanup: true,
+			},
+		},
+	}
+
+	stepDetachOldOSDisk, err := daisyutils.NewStep(w, "detach-old-os-disk", stepCreateNewOSDisk)
+	if err != nil {
+		return err
+	}
+	stepDetachOldOSDisk.DetachDisks = &daisy.DetachDisks{
+		&daisy.DetachDisk{
+			Instance:   u.InstanceURI,
+			DeviceName: daisyutils.GetDeviceURI(u.project, u.zone, u.osDiskDeviceName),
+		},
+	}
+
+	stepAttachNewOSDisk, err := daisyutils.NewStep(w, "attach-new-os-disk", stepDetachOldOSDisk)
+	if err != nil {
+		return err
+	}
+	stepAttachNewOSDisk.AttachDisks = &daisy.AttachDisks{
+		&daisy.AttachDisk{
+			Instance: u.InstanceURI,
+			AttachedDisk: compute.AttachedDisk{
+				Source:     u.newOSDiskName,
+				DeviceName: u.osDiskDeviceName,
+				AutoDelete: u.osDiskAutoDelete,
+				Boot:       true,
+			},
+		},
+	}
+
+	stepCreateInstallDisk, err := daisyutils.NewStep(w, "create-install-disk", stepAttachNewOSDisk)
+	if err != nil {
+		return err
+	}
+	stepCreateInstallDisk.CreateDisks = &daisy.CreateDisks{
+		&daisy.Disk{
+			Disk: compute.Disk{
+				Name:        u.installMediaDiskName,
+				Zone:        u.zone,
+				Type:        "pd-ssd",
+				SourceImage: "projects/compute-image-tools/global/images/family/windows-install-media",
+			},
+			Resource: daisy.Resource{
+				ExactName: true,
+				NoCleanup: true,
+			},
+		},
+	}
+
+	stepAttachInstallDisk, err := daisyutils.NewStep(w, "attach-install-disk", stepCreateInstallDisk)
+	if err != nil {
+		return err
+	}
+	stepAttachInstallDisk.AttachDisks = &daisy.AttachDisks{
+		&daisy.AttachDisk{
+			Instance: u.InstanceURI,
+			AttachedDisk: compute.AttachedDisk{
+				Source:     u.installMediaDiskName,
+				AutoDelete: true,
+			},
+		},
+	}
+	prevStep = stepAttachInstallDisk
+
+	// 'windows-startup-script-url' exists, backup it
```

Comment:
Can you expand the comment explaining your backup tactic

Reply:
I explained details in validators.go line 165 in the latest commit. I  the latest commit, the logic here is not complex anymore.